### PR TITLE
Implementation of `CryptographicSponge` for Merlin

### PR DESCRIPTION
**Pull request by autquis**
Using Merlin, we have the possibility of having a sponge that is not generic on a field.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (main)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -25,6 +25,7 @@ ark-serialize = { version = "^0.4.0", default-features = false, features = [ "de
 blake2 = { version = "0.10", default-features = false }
 sha2 = { version = "0.10", default-features = false }
 digest = { version = "0.10", default-features = false }
+merlin = { version = "3.0.0", default-features = false }
```
Comment (mmagician) [inline Cargo.toml:28]: We can make this an optional dependency, enabled under the `sponge` feature

File: src/sponge/merlin/mod.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+use crate::sponge::{Absorb, CryptographicSponge};
```
Comment (mmagician) [inline src/sponge/merlin/mod.rs:1]:
```suggestion
use crate::sponge::{Absorb, CryptographicSponge};
use crate::Vec;
```

File: src/sponge/merlin/mod.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+use merlin::Transcript;
```
Comment (mmagician) [inline src/sponge/merlin/mod.rs:2]:
```suggestion
pub use merlin::Transcript;
```

**Comment by mmagician** [comment]
@hdevalence

File: crypto-primitives/src/sponge/mod.rs [hunk 4]
```diff
@@ -17,6 +17,11 @@ pub use absorb::*;
 /// [cos]: https://eprint.iacr.org/2019/1076
 pub mod poseidon;
 
+/// The sponge for Merlin
+///
+///
```
Comment (mmagician) [inline crypto-primitives/src/sponge/mod.rs:22]: Is this perhaps missing a link/description?

Comment (mmagician) [reply crypto-primitives/src/sponge/mod.rs:22]: @autquis

Comment (autquis) [reply crypto-primitives/src/sponge/mod.rs:22]: Sorry, it slipped out of my hands!

**Comment by autquis** [comment]
I had to remove the redundant imports since `cargo +nightly check` complained. Specifically, `Vec` is in the prelude and is redundant to import.

**Comment by mmagician** [comment]
@autquis ah yes see also this PR: https://github.com/arkworks-rs/algebra/pull/790 to tackle the same sort of issues in algebra

**Comment by autquis** [comment]
@mmagician I didn't understand what I should fix in arkworks-rs/algebra#790? As it is merged already.

Why it works: I guess because it is not completely `no-default-features` and `r1cs` is enabled. This is the command for no_std check: 
```
cargo build --no-default-features --features=r1cs --target aarch64-unknown-none
```

**Comment by mmagician** [comment]
What I meant is that the algebra PR addressed similar issues.

**Comment by autquis** [comment]
A gentle ping :)

**Comment by mmagician** [comment]
Also @mmaker since I'm tagging you in a few places now, maybe you could also take a peek at this?

**Comment by autquis** [comment]
CI failure is due to this change probably rust-lang/rust#121752
@Cesar199999 @antonio95

**Comment by Cesar199999** [comment]
@autquis Yes sorry, I had to update the branch, I'll fix it

**Comment by mmaker** [comment]
Took a quick peek! The PR looks okay, but there isn't much to say. I'd personally encourage your to add a few lines of documentation so that people are well-aware of the limitations of this trait, especially concerning the statement serialization (the user is on their own) and the bit vs bytes part. 

That said, I'm happy to see that people are starting to need something that works both over bytes and algebraic sponges :)

**Review by Pratyush — APPROVED** [review]

**Pull request merged**